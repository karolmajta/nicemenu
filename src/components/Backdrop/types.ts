import { ReactChild } from "react";

export interface BackdropProps {
  children: ReactChild;
  className?: string;
  classNames?: Array<string>;
  onDismiss: () => void;
}
