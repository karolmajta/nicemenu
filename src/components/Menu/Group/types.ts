import { IconName } from "../Item/Icon/types";
import React from "react";
import { AllowedShortcutKey, ItemProps } from "../Item/types";

export interface GroupItemProps {
  icon: IconName;
  label: string;
  shortcut?: Array<Array<AllowedShortcutKey>>;
  goto?: string;
  onSelect?: () => void;
  listIndex: number;
}

export interface GroupProps {
  label: string;
  items: Array<GroupItemProps>;
  activeListIndex: number | null;
  className?: string;
  classNames?: Array<string>;
  components?: {
    Item?: React.ComponentType<ItemProps>;
  };
  onItemMouseMove?: (index: number) => void;
  onGoto: (gotoPath: string) => void;
}
