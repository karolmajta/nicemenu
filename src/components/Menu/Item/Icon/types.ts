export type IconName =
  | "bolt"
  | "dot"
  | "inbox"
  | "loader"
  | "pencil"
  | "reschedule";

export interface IconProps {
  name: IconName;
  className?: string;
  classNames?: Array<string>;
}
