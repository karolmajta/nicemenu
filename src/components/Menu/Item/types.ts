import React from "react";
import { IconName, IconProps } from "./Icon/types";

export type AllowedShortcutKey =
  | "a"
  | "b"
  | "c"
  | "d"
  | "e"
  | "f"
  | "g"
  | "h"
  | "i"
  | "j"
  | "k"
  | "l"
  | "m"
  | "n"
  | "o"
  | "p"
  | "q"
  | "r"
  | "s"
  | "t"
  | "u"
  | "v"
  | "w"
  | "x"
  | "y"
  | "z"
  | "Shift";

export interface ItemProps {
  label: string;
  icon?: IconName;
  shortcut?: Array<Array<AllowedShortcutKey>>;
  listIndex: number;
  highlighted: boolean;
  goto?: string;
  className?: string;
  classNames?: Array<string>;
  components?: {
    Icon?: React.ComponentType<IconProps>;
  };
  onMouseMove?: (listIndex: number) => void;
  onSelect?: () => void;
  onGoto: (menuKey: string) => void;
}
