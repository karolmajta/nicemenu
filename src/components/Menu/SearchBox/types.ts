export interface SearchBoxProps {
  placeholder: string;
  searchTerm: string;
  focused: boolean;
  className?: string;
  classNames?: Array<string>;
  onArrowUp: () => void;
  onArrowDown: () => void;
  onChange: (searchTerm: string) => void;
  onFocus: () => void;
}
