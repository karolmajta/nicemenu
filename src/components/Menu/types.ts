import { IconName } from "./Item/Icon/types";
import React from "react";
import { ItemProps, AllowedShortcutKey } from "./Item/types";
import { GroupProps } from "./Group/types";
import { SearchBoxProps } from "./SearchBox/types";

export interface MenuItem {
  icon: IconName;
  label: string;
  shortcut?: Array<Array<AllowedShortcutKey>>;
  goto?: string;
  onSelect?: () => void;
}

export interface MenuGroup {
  label: string;
  items: Array<MenuItem>;
}

export interface MenuProps {
  title: string;
  items: Array<MenuItem>;
  groups: Array<MenuGroup>;
  submenus: {
    [submenuName: string]: MenuProps;
  };
  className?: string;
  classNames?: Array<string>;
  components?: {
    Item?: React.ComponentType<ItemProps>;
    Group?: React.ComponentType<GroupProps>;
    SearchBox?: React.ComponentType<SearchBoxProps>;
  };
}
