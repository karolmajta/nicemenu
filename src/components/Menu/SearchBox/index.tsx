import React, { useRef, useEffect } from "react";
import cx from "classnames";
import styled from "styled-components";

import { SearchBoxProps } from "./types";

export function SearchBoxUnstyled({
  placeholder,
  searchTerm,
  className,
  classNames,
  focused,
  onArrowUp,
  onArrowDown,
  onChange,
  onFocus,
}: SearchBoxProps) {
  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    if (inputRef.current) {
      if (focused) {
        inputRef.current.focus();
        onChange("");
      } else {
        inputRef.current.blur();
      }
    }
  }, [focused, onChange]);

  return (
    <div
      data-testid="Laserfocus-Menu-SearchBox"
      className={cx(
        "Laserfocus-Menu-SearchBox",
        className,
        ...(classNames || [])
      )}
    >
      <input
        type="text"
        placeholder={placeholder}
        value={searchTerm}
        onChange={(e) => onChange(e.target.value)}
        onKeyDown={(e) => {
          e.stopPropagation();
          e.key === "ArrowUp" && onArrowUp();
          e.key === "ArrowDown" && onArrowDown();
        }}
        onKeyUp={(e) => {
          if (e.key !== "Escape") {
            e.stopPropagation();
          }
        }}
        onFocus={onFocus}
        ref={inputRef}
      />
    </div>
  );
}

const SearchBox = styled(SearchBoxUnstyled)`
  display: flex;
  justify-content: center;
  align-items: center;

  width: 100%;
  border-bottom: 1px solid rgba(228, 228, 238, 0.2);

  input {
    display: flex;
    margin: 12px 20px;
    width: calc(100% - 2 * 20px);
    height: 24px;

    background: none;
    border: none;
    outline: none;
    caret-color: #176397;
    font-family: Inter;
    font-size: 20px;
    line-height: 24px;
    color: rgba(228, 228, 238, 0.5);

    &::placeholder {
      color: rgba(228, 228, 238, 0.2);
    }
  }
`;

export default SearchBox;
