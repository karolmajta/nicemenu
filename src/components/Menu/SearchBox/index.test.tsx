import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import SearchBox from ".";
import { SearchBoxProps } from "./types";

const renderSearchBox = (props: Partial<SearchBoxProps>) =>
  render(
    <SearchBox
      placeholder={props.placeholder || "Placeholder"}
      searchTerm={props.searchTerm || ""}
      focused={props.focused || false}
      onArrowUp={props.onArrowUp || jest.fn()}
      onArrowDown={props.onArrowDown || jest.fn()}
      onChange={props.onChange || jest.fn()}
      onFocus={props.onFocus || jest.fn()}
    />
  );

afterEach(() => {
  cleanup();
});

test("renders correctly", () => {
  const { getByTestId } = renderSearchBox({});
  const searchBoxDiv = getByTestId("Laserfocus-Menu-SearchBox");
  expect(searchBoxDiv).toMatchSnapshot();
});

test("renders with input focused if focused is true", () => {
  const { getByTestId } = renderSearchBox({ focused: true });
  const searchBoxDiv = getByTestId("Laserfocus-Menu-SearchBox");
  const input = searchBoxDiv.querySelector("input") as HTMLInputElement;
  expect(window.document.activeElement).toBe(input);
});

test("calls onArrowUp correctly", () => {
  const onArrowUp = jest.fn();
  const { getByTestId } = renderSearchBox({ onArrowUp });
  const searchBoxDiv = getByTestId("Laserfocus-Menu-SearchBox");
  const input = searchBoxDiv.querySelector("input") as HTMLInputElement;
  fireEvent.keyDown(input, { key: "ArrowUp" });
  expect(onArrowUp).toHaveBeenCalled();
});

test("calls onArrowDown correctly", () => {
  const onArrowDown = jest.fn();
  const { getByTestId } = renderSearchBox({ onArrowDown });
  const searchBoxDiv = getByTestId("Laserfocus-Menu-SearchBox");
  const input = searchBoxDiv.querySelector("input") as HTMLInputElement;
  fireEvent.keyDown(input, { key: "ArrowDown" });
  expect(onArrowDown).toHaveBeenCalled();
});

test("calls onChange correctly", () => {
  const onChange = jest.fn();
  const { getByTestId } = renderSearchBox({ onChange });
  const searchBoxDiv = getByTestId("Laserfocus-Menu-SearchBox");
  const input = searchBoxDiv.querySelector("input") as HTMLInputElement;
  fireEvent.change(input, { target: { value: "Hello" } });
  expect(onChange).toHaveBeenCalledWith("Hello");
});

test("calls onFocus correctly", () => {
  const onFocus = jest.fn();
  const { getByTestId } = renderSearchBox({ onFocus });
  const searchBoxDiv = getByTestId("Laserfocus-Menu-SearchBox");
  const input = searchBoxDiv.querySelector("input") as HTMLInputElement;
  input.focus();
  expect(onFocus).toHaveBeenCalled();
});
