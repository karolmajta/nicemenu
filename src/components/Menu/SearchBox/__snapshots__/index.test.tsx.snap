// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  class="Laserfocus-Menu-SearchBox sc-bdnylx cCWaaj"
  data-testid="Laserfocus-Menu-SearchBox"
>
  <input
    placeholder="Placeholder"
    type="text"
    value=""
  />
</div>
`;
